File: app/src/main/java/io/github/sithengineer/motoqueiro/ui/cruising/CruisingPresenter.java
package io.github.sithengineer.motoqueiro.ui.cruising;

import io.github.sithengineer.motoqueiro.util.CompositeSubscriptionManager;
import timber.log.Timber;

// TODO
// insert data capture in batches
// use an entity to handle data save / load logic, given a repository
public class CruisingPresenter implements CruisingContract.Presenter {

  private final CruisingContract.View view;
  private final CompositeSubscriptionManager subscriptionManager;
  private final CruisingNavigator navigator;

  public CruisingPresenter(CruisingContract.View view,
      CompositeSubscriptionManager subscriptionManager, CruisingNavigator navigator) {
    this.view = view;
    this.subscriptionManager = subscriptionManager;
    this.navigator = navigator;
  }

  @Override public void start() {
    handleStopClick();
  }

  @Override public void stop() {
    subscriptionManager.clearAll();
  }

  private void handleStopClick() {
    subscriptionManager.add(
        view.stopClick()
            .doOnNext(__2 -> view.setStopButtonDisabled())
            .doOnNext(__2 -> navigator.stopServiceToGatherData())
            .doOnNext(__2 -> navigator.goToStatistics())
            .retry()
        .subscribe(__ -> {
        }, err -> Timber.e(err)));
  }
}

File: app/src/main/java/io/github/sithengineer/motoqueiro/ui/cruising/CruisingPresenter.kt
package io.github.sithengineer.motoqueiro.ui.cruising

import io.github.sithengineer.motoqueiro.util.CompositeSubscriptionManager
import io.reactivex.rxkotlin.subscribeBy
import timber.log.Timber

// TODO
// insert data capture in batches
// use an entity to handle data save / load logic, given a repository
class CruisingPresenter(private val view: CruisingContract.View,
    private val subscriptionManager: CompositeSubscriptionManager,
    private val navigator: CruisingNavigator) : CruisingContract.Presenter {

  override fun start() {
    handleStopClick()
  }

  override fun stop() {
    subscriptionManager.clearAll()
  }

  private fun handleStopClick() {
    subscriptionManager.add(
        view.stopClick()
            .doOnNext { view.setStopButtonDisabled() }
            .doOnNext { navigator.stopServiceToGatherData() }
            .doOnNext { navigator.goToStatistics() }
            .retry()
            .subscribeBy(
                onComplete = { },
                onError = { err -> Timber.e(err) }
            )
    )
  }
}
